src/Workspaces/Core/Portable/Simplification/AbstractSimplificationService.cs

```diff
@@ -178,89 +178,92 @@ protected virtual SyntaxNode TransformReducedNode(SyntaxNode reducedNode, Syntax
-                var isNode = nodeOrToken.IsNode;
-
-                foreach (var reducer in reducers)
                 {
-                    cancellationToken.ThrowIfCancellationRequested();
+                    var nodeOrToken = nodeOrTokenToReduce.OriginalNodeOrToken;
+                    var simplifyAllDescendants = nodeOrTokenToReduce.SimplifyAllDescendants;
+                    var semanticModelForReduce = semanticModel;
+                    var currentNodeOrToken = nodeOrTokenToReduce.NodeOrToken;
+                    var isNode = nodeOrToken.IsNode;
 
-                    var rewriter = reducer.CreateExpressionRewriter(optionSet, cancellationToken);
-
-                    do
+                    foreach (var reducer in reducers)
                     {
-                        if (currentNodeOrToken.SyntaxTree != semanticModelForReduce.SyntaxTree)
-                        {
-                            // currentNodeOrToken was simplified either by a previous reducer or
-                            // a previous iteration of the current reducer.
-                            // Create a speculative semantic model for the simplified node for semantic queries.
-
-                            // Certain node kinds (expressions/statements) require non-null parent nodes during simplification.
-                            // However, the reduced nodes haven't been parented yet, so do the required parenting using the original node's parent.
-                            if (currentNodeOrToken.Parent == null &&
-                                nodeOrToken.Parent != null &&
-                                (currentNodeOrToken.IsToken ||
-                                currentNodeOrToken.AsNode() is TExpressionSyntax ||
-                                currentNodeOrToken.AsNode() is TStatementSyntax ||
-                                currentNodeOrToken.AsNode() is TCrefSyntax))
-                            {
-                                var annotation = new SyntaxAnnotation();
-                                currentNodeOrToken = currentNodeOrToken.WithAdditionalAnnotations(annotation);
+                        cancellationToken.ThrowIfCancellationRequested();
 
-                                var replacedParent = isNode ?
-                                    nodeOrToken.Parent.ReplaceNode(nodeOrToken.AsNode(), currentNodeOrToken.AsNode()) :
-                                    nodeOrToken.Parent.ReplaceToken(nodeOrToken.AsToken(), currentNodeOrToken.AsToken());
-
-                                currentNodeOrToken = replacedParent
-                                    .ChildNodesAndTokens()
-                                    .Single((c) => c.HasAnnotation(annotation));
-                            }
+                        using (var rewriter = reducer.GetOrCreateRewriter())
+                        {
+                            rewriter.Initialize(optionSet, cancellationToken);
```

Comment:
these three lines are new.  We now get a potentially pooled rewriter, and we initialize it with the current state.